File: allcode/3200-3299/3201maximumLength.py


from leetcode.allcode.competition.mypackage import *

class Solution:
    def maximumLength(self, nums: List[int]) -> int:
        odd = even = 0
        for x in nums:
            if x & 1:
                odd = max(odd, even + 1)
            else:
                even = max(even, odd + 1)
        nodd = sum(1 for x in nums if x & 1)
        return max(odd, even, nodd, len(nums) - nodd, 2)


so = Solution()
print(so.maximumLength([4,2,6]))
print(so.maximumLength([1,2,3,4]))
print(so.maximumLength([1,2,1,1,2,1,2]))
print(so.maximumLength([1,3]))

File: c++/src/3000-3999/3201maximumLength.cpp
#include "lc_pub.h"


class Solution {
    public:
    int maximumLength(vector<int>& nums) {
        int dp0=0,dp1=0;
        int n=nums.size();
        int odd;
        if (nums[0]&1) {dp1=1;odd=1;}
        else dp0=1;
        for (int i=1;i<n;i++) {
            if (nums[i]&1) {
                dp1=dp0+1;
                odd++;
            }
            else {
                dp0=dp1+1;
            }
        }
        int ans=max(odd,n-odd);
        return max(ans, max(dp0,dp1));
    }
    };

    
int main()
{
    cout<<"test let us start! %s" << __cplusplus <<std::endl;
    vector<int> nums{1,2,3,4};

    Solution so;
    cout << so.maximumLength(nums) << endl;
    return 0;
}
